### PR TITLE
Add quoted.Expr.asExprOf and Reflection.Tree.{asExprOf, isExpr}

### PR DESCRIPTION
**Pull request by nicolasstucki**
These methods allow checked casts from `Expr` or a `Tree`.
`Tree.asExprOf` is similar to `Trem.seal` with the difference that it also checks if the tree is a valid term.

Add `isExpr` and improve implemetation of `seal`, `sealOpt` and `ExprMap`.

### CODE REVIEW

File: library/src-bootstrapped/scala/quoted/Expr.scala [hunk 1]
```diff
@@ -41,11 +41,15 @@ abstract class Expr[+T] private[scala] {
...
-  def cast[U](using tp: scala.quoted.Type[U])(using qctx: QuoteContext): scala.quoted.Expr[U] = {
+  def cast[U](using tp: scala.quoted.Type[U])(using qctx: QuoteContext): scala.quoted.Expr[U] = asExprOf[U]
+
+  /** Convert this to an `quoted.Expr[X]` if this expression is a valid expression of type `X` or throws */
+  def asExprOf[X](using tp: scala.quoted.Type[X])(using qctx: QuoteContext): scala.quoted.Expr[X] = {
     val tree = this.unseal
     val expectedType = tp.unseal.tpe
     if (tree.tpe <:< expectedType)
-      this.asInstanceOf[scala.quoted.Expr[U]]
+      this.asInstanceOf[scala.quoted.Expr[T]]
```
Comment (liufengyun) [inline library/src-bootstrapped/scala/quoted/Expr.scala:51]:
```suggestion
```

File: library/src/scala/tasty/Reflection.scala [hunk 2]
```diff
@@ -488,7 +488,28 @@ class Reflection(private[scala] val internal: CompilerInterface) { self =>
...
+
+      /** Does this tree represent a valid expression? */
+      def isExpr(using ctx: Context): Boolean =
+        tree match
+          case tree: Term =>
+            tree.tpe.widen match
+              case _: MethodType | _: PolyType => false
+              case _ => true
+          case _ => false
+
```
Comment (liufengyun) [inline library/src/scala/tasty/Reflection.scala:500]: The `isExpr` will cause confusion for meta-programmers: what's the relation between `expression` and `Expr[T]`? What is the relationship between `expression` and `Term`?

Comment (nicolasstucki) [reply library/src/scala/tasty/Reflection.scala:500]: The is `isExpr` indicates that this term is a valid `Expr[Any]`, therefore `asExprOf` will succeed. The names are intentionally similar as the are about the same concept. Docs can improve.

Comment (liufengyun) [reply library/src/scala/tasty/Reflection.scala:500]: Maybe make the method private?

Comment (nicolasstucki) [reply library/src/scala/tasty/Reflection.scala:500]: But users are intended to use it

**Pull request merged**